pykern/pkio.py

```diff
@@ -125,6 +125,30 @@ def has_file_extension(filename, to_check):
     return e in to_check
 
 
+def is_binary(filename, threshold=0.30):
+    """Uses heuristics to guess whether file is binary data
+
+    Args:
+        filename (str|py.path.local): file to check
+        threshold (float): allowed percentage of assumed non-textual characters
+
+    Returns:
+        bool: True if file is likely binary data
+    """
+    if threshold >= 1:
+        raise AssertionError(
+            f"threshold={threshold} must be < 1 (represents percentage)"
+        )
+    with open(filename, "rb") as f:
+        b = f.read(512)
+        if not b:
+            return False
+        if b"\x00" in b:
+            return True
+        l = len(b)
+        return (l - len(b.decode("utf-8", "ignore"))) / l > threshold
```

Comment:
If it can't decode utf8, then it should be binary, i.e. do not ignore errors. The likelihood of truncating at a utf8 boundary is low but real. Try to decode as utf8. If succeeds, then it's text. If it fails, look at the last three bytes to see if there's a utf start character and backup from there, possibly iteratively. 

I am not clear the threshold applies any more now that we have utf8 almost exclusively. We don't have to worry about other codecs. It may be that simply decoding for utf8 is sufficient.

We might add an iso 8859 codec, since that would be easy. I don't think we need to handle much else.